File: rps.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>

using namespace std;

enum class Choice { ROCK, PAPER, SCISSORS };

// Function to convert choice from int to enum
Choice intToChoice(int choiceInt) {
    switch (choiceInt) {
        case 0:
            return Choice::ROCK;
        case 1:
            return Choice::PAPER;
        case 2:
            return Choice::SCISSORS;
        default:
            return Choice::ROCK;
    }
}

// Function to print enum choice as string
string choiceToString(Choice choice) {
    switch (choice) {
        case Choice::ROCK:
            return "Rock";
        case Choice::PAPER:
            return "Paper";
        case Choice::SCISSORS:
            return "Scissors";
        default:
            return "Unknown";
    }
}

// Function to get player's choice
Choice getPlayerChoice() {
    int choiceInt;
    cout << "Enter your choice (0 - Rock, 1 - Paper, 2 - Scissors): ";
    cin >> choiceInt;
    return intToChoice(choiceInt);
}

// Function to get bot's choice
Choice getBotChoice() {
    int choiceInt = rand() % 3;
    return intToChoice(choiceInt);
}

// Function to determine the winner
int determineWinner(Choice playerChoice, Choice botChoice) {
    if (playerChoice == botChoice) {
        return 0;  // Draw
    } else if ((playerChoice == Choice::ROCK && botChoice == Choice::SCISSORS) ||
               (playerChoice == Choice::PAPER && botChoice == Choice::ROCK) ||
               (playerChoice == Choice::SCISSORS && botChoice == Choice::PAPER)) {
        return 1;  // Player wins
    } else {
        return 2;  // Bot wins
    }
}

int main() {
    // Seed the random number generator
    srand(time(0));

    int gameMode;
    cout << "Welcome to Rock-Paper-Scissors game!" << endl;
    cout << "Select a game mode (1 - Player vs Player, 2 - Player vs Bot, 3 - Bot vs Bot): ";
    cin >> gameMode;

    Choice player1Choice, player2Choice;
    int winner;

    switch (gameMode) {
        case 1:
            // Player vs Player mode
            player1Choice = getPlayerChoice();
            player2Choice = getPlayerChoice();
            winner = determineWinner(player1Choice, player2Choice);
            break;
        case 2:
            // Player vs Bot mode
            player1Choice = getPlayerChoice();
            player2Choice = getBotChoice();
            winner = determineWinner(player1Choice, player2Choice);
            break;
        case 3:
            // Bot vs Bot mode
            player1Choice = getBotChoice();
            player2Choice = getBotChoice();
            winner = determineWinner(player1Choice, player2Choice);
            break;
        default:
            cout << "Invalid game mode. Exiting..." << endl;
            return 0;
    }

    // Print the results
    cout << "Player 1 choice: " << choiceToString(player1Choice) << endl;
    cout << "Player 2 choice: " << choiceToString(player2Choice) << endl;
    if (winner == 0) {
        cout << "It's a draw!" << endl;
    } else if (winner == 1) {
cout << "Player 1 wins!" << endl;
} else {
cout << "Player 2 wins!" << endl;
}
return 0;
}

File: rps.py
import random

MODE_PLAYER_VS_PLAYER = 1
MODE_PLAYER_VS_BOT = 2
MODE_BOT_VS_BOT = 3

CHOICES = ['rock', 'paper', 'scissors']

scoreboard = {'Player 1': {'wins': 0, 'draws': 0, 'losses': 0},
              'Player 2': {'wins': 0, 'draws': 0, 'losses': 0}}


def get_player_choice():
    while True:
        print("Enter your choice (rock, paper, scissors):")
        choice = input().lower()
        if choice in CHOICES:
            return choice
        else:
            print("Invalid choice. Please try again.")


def get_bot_choice():
    return random.choice(CHOICES)


def determine_winner(player1_choice, player2_choice):
    if player1_choice == player2_choice:
        return 'draw'
    elif (player1_choice == 'rock' and player2_choice == 'scissors') or \
            (player1_choice == 'paper' and player2_choice == 'rock') or \
            (player1_choice == 'scissors' and player2_choice == 'paper'):
        return 'Player 1'
    else:
        return 'Player 2'


def display_scoreboard():
    print("\nScoreboard")
    print("Player 1: Wins - {}, Draws - {}, Losses - {}".format(scoreboard['Player 1']['wins'],
                                                                scoreboard['Player 1']['draws'],
                                                                scoreboard['Player 1']['losses']))
    print("Player 2: Wins - {}, Draws - {}, Losses - {}".format(scoreboard['Player 2']['wins'],
                                                                scoreboard['Player 2']['draws'],
                                                                scoreboard['Player 2']['losses']))


def reset_scoreboard():
    scoreboard['Player 1']['wins'] = 0
    scoreboard['Player 1']['draws'] = 0
    scoreboard['Player 1']['losses'] = 0
    scoreboard['Player 2']['wins'] = 0
    scoreboard['Player 2']['draws'] = 0
    scoreboard['Player 2']['losses'] = 0


def play_game():
    while True:
        print("\nSelect game mode:")
        print("1. Player vs Player")
        print("2. Player vs Bot")
        print("3. Bot vs Bot")
        print("4. Quit")
        mode = input()

        if mode == '1':  # Player vs Player
            print("\nPlayer 1:")
            player1_choice = get_player_choice()
            print("\nPlayer 2:")
            player2_choice = get_player_choice()

        elif mode == '2':  # Player vs Bot
            print("\nPlayer 1:")
            player1_choice = get_player_choice()
            player2_choice = get_bot_choice()
            print("Player 2 chooses: " + player2_choice)

        elif mode == '3':  # Bot vs Bot
            player1_choice = get_bot_choice()
            player2_choice = get_bot_choice()
            print("Player 1 chooses: " + player1_choice)
            print("Player 2 chooses: " + player2_choice)

        elif mode == '4':  # Quit
            print("Thank you for playing!")
            break

        else:
            print("Invalid choice. Please try again.")
            continue

        winner = determine_winner(player1_choice, player2_choice)

        if winner == 'draw':
            print("It's a draw!")
            scoreboard['Player 1']['draws']
        elif winner == 'Player 1':
            print("Player 1 wins!")
            scoreboard['Player 1']['wins'] += 1
            scoreboard['Player 2']['losses'] += 1
        else:
            print("Player 2 wins!")
            scoreboard['Player 1']['losses'] += 1
            scoreboard['Player 2']['wins'] += 1

        display_scoreboard()

        # Check if any player has reached 3 wins
        if scoreboard['Player 1']['wins'] == 3 or scoreboard['Player 2']['wins'] == 3:
            print("Game over!")
            if scoreboard['Player 1']['wins'] == 3:
                print("Player 1 wins the game!")
            else:
                print("Player 2 wins the game!")
            reset_scoreboard()
            continue

        # Ask if the player wants to play another round
        print("\nDo you want to play another round? (yes/no)")
        play_again = input().lower()
        if play_again != 'yes':
            print("Thank you for playing!")
            break


if __name__ == '__main__':
    play_game()
